### PR TITLE
Polyhedron_demo: Enhance memory management in the demo

### LINKED ISSUE #1957: Polyhedron Demo: Memory leak in the off_plugin or Scene_polyhedron_item
## Issue Details

There seems to be an important memory leak in the Polyhedron Demo, while loading/erasing polyhedron items.

That should be fixed as soon as possible.

## Source Code

The following command line loads a Qt Script that loads/erases an OFF file 500 times. The memory usage of the application increases, even if the items are destroyed, up to 7.0GB on my machine.

```shell
./Polyhedron_3 qtscript:'for(i=0; i< 500; ++i) { main_window.open("data/man.off", "off_plugin"); scene.erase(scene.item(0)); print(i);}'
```

I have chosen `data/man.off` for the OFF file. Adapt the path to it on your machine.

## Environment

* Operating system (Windows/Mac/Linux, 32/64 bits): **Linux 64 bits, Fedora 24**
* Compiler:
* Release or debug mode:
* Specific flags used (if any):
* CGAL version: **master**, commit 2e7d96f87e88f5950e24d14bede00a87ef646e74
* Boost version:
* Other libraries versions if used (Eigen, TBB, etc.):

### PR DESCRIPTION
**Pull request by maxGimeno**
## Summary of Changes

Add a function to Scene_item that allows to delete things when the signal aboutToBeDestroyed is called, and use it on the big-data items so the memory is freed when it must be.

## Release Management

* Affected package(s): Polyhedron_demo
* Issue(s) solved (if any): fix #1957

### CODE REVIEW

**Comment by sloriot** [comment]
@lrineau  shall this target 4.10?

**Comment by lrineau** [comment]
Yes, that should be in CGAL-4.10-beta.

**Comment by lrineau** [comment]
@maxGimeno  I suggest you rebase on `master`, to resolve the conflict.

File: Polyhedron/demo/Polyhedron/Scene_c3t3_item.cpp [hunk 1]
```diff
@@ -281,10 +281,19 @@ struct Scene_c3t3_item_priv {
...
+    c3t3.clear();
+    tree.clear();
+    if(frame)
+    {
+      static_cast<CGAL::Three::Viewer_interface*>(QGLViewer::QGLViewerPool().first())->setManipulatedFrame(0);
+      delete frame;
+      frame = NULL;
+      delete tet_Slider;
+    }
+  }
```
Comment (lrineau) [inline Polyhedron/demo/Polyhedron/Scene_c3t3_item.cpp]: I do no understand why there is a function `Scene_c3t3_item_priv::destroy()`. Why is not it merged with `~Scene_c3t3_priv()`?

File: Polyhedron/demo/Polyhedron/Scene_c3t3_item.h [hunk 2]
```diff
@@ -149,6 +149,8 @@ using namespace CGAL::Three;
 
   QColor get_histogram_color(const double v) const;
 
+  void itemAboutToBeDestroyed(Scene_item *);
```
Comment (lrineau) [inline Polyhedron/demo/Polyhedron/Scene_c3t3_item.h]: You should add the `Q_OVERRIDE` macro to it (I have not verified the precise name).

File: Polyhedron/demo/Polyhedron/Scene_c3t3_item.cpp [hunk 3]
```diff
@@ -1988,6 +2009,22 @@ void Scene_c3t3_item::invalidateOpenGLBuffers()
...
+  if(!d)
+    return;
  d->reset_cut_plane();
 }
+
+void Scene_c3t3_item::itemAboutToBeDestroyed(Scene_item *item)
+{
+  Scene_item::itemAboutToBeDestroyed(item);
+
+  if(d)
```
Comment (lrineau) [inline Polyhedron/demo/Polyhedron/Scene_c3t3_item.cpp]: `if(d && item == this)`

File: Polyhedron/demo/Polyhedron/Scene_points_with_normal_item.cpp [hunk 4]
```diff
@@ -870,3 +873,13 @@ int Scene_points_with_normal_item::getPointSliderValue()
 {
   return d->point_Slider->value();
 }
+
+void Scene_points_with_normal_item::itemAboutToBeDestroyed(Scene_item *item)
+{
+  Scene_item::itemAboutToBeDestroyed(item);
+  if(d && d->m_points)
```
Comment (lrineau) [inline Polyhedron/demo/Polyhedron/Scene_points_with_normal_item.cpp]: `&& item == this`

File: Polyhedron/demo/Polyhedron/Scene_polygon_soup_item.cpp [hunk 5]
```diff
@@ -846,3 +850,17 @@ const Scene_polygon_soup_item::Points& Scene_polygon_soup_item::points() const {
 bool Scene_polygon_soup_item::isDataColored() { return d->soup->fcolors.size()>0 || d->soup->vcolors.size()>0;}
 std::vector<CGAL::Color> Scene_polygon_soup_item::getVColors() const{return d->soup->vcolors;}
 std::vector<CGAL::Color> Scene_polygon_soup_item::getFColors() const{return d->soup->fcolors;}
+
+void Scene_polygon_soup_item::itemAboutToBeDestroyed(Scene_item *item)
+{
+  Scene_item::itemAboutToBeDestroyed(item);
+ if(d)
```
Comment (lrineau) [inline Polyhedron/demo/Polyhedron/Scene_polygon_soup_item.cpp]: `d && item == this`

File: Polyhedron/demo/Polyhedron/Scene_polyhedron_item.cpp [hunk 6]
```diff
@@ -892,14 +892,22 @@ Scene_polyhedron_item::~Scene_polyhedron_item()
...
+        v->textRenderer->removeTextList(textItems);
+        delete textItems;
+        textItems=NULL;
+      }
+    }
+    if(d)
+    {
+      d->destroy();
+      delete d;
+      d=NULL;
```
Comment (lrineau) [inline Polyhedron/demo/Polyhedron/Scene_polyhedron_item.cpp]: Same here: why should one call `destroy()` before the destructor?

File: Polyhedron/demo/Polyhedron/Scene_polyhedron_item.cpp [hunk 7]
```diff
@@ -1821,3 +1829,15 @@ bool Scene_polyhedron_item::intersect_face(double orig_x,
   return false;
 
 }
+
+void Scene_polyhedron_item::itemAboutToBeDestroyed(Scene_item *item)
+{
+  Scene_item::itemAboutToBeDestroyed(item);
+  if(d)
```
Comment (lrineau) [inline Polyhedron/demo/Polyhedron/Scene_polyhedron_item.cpp]: `&& this ==item`

File: Polyhedron/demo/Polyhedron/Scene_polyhedron_item.h [hunk 8]
```diff
@@ -132,6 +132,7 @@ public Q_SLOTS:
     void update_facet_indices();
     void update_halfedge_indices();
     void invalidate_aabb_tree();
+    void itemAboutToBeDestroyed(Scene_item *);
```
Comment (lrineau) [inline Polyhedron/demo/Polyhedron/Scene_polyhedron_item.h]: `Q_OVERRIDE`

File: Polyhedron/demo/Polyhedron/Scene_surface_mesh_item.cpp [hunk 9]
```diff
@@ -623,3 +627,12 @@ void Scene_surface_mesh_item::compute_bbox()const
 
 }
 
+void Scene_surface_mesh_item::itemAboutToBeDestroyed(Scene_item *item)
+{
+  Scene_item::itemAboutToBeDestroyed(item);
+  if(d && d->smesh_)
```
Comment (lrineau) [inline Polyhedron/demo/Polyhedron/Scene_surface_mesh_item.cpp]: `&& this == item`

File: Polyhedron/demo/Polyhedron/Scene_polygon_soup_item.cpp [hunk 10]
```diff
@@ -846,3 +850,17 @@ const Scene_polygon_soup_item::Points& Scene_polygon_soup_item::points() const {
...
+{
+  Scene_item::itemAboutToBeDestroyed(item);
+ if(d && item == this)
+ {
+   if(d->soup)
+   {
+     delete d->soup;
+     d->soup=NULL;
+   }
+ }
```
Comment (lrineau) [inline Polyhedron/demo/Polyhedron/Scene_polygon_soup_item.cpp]: That's a lot better. All I can find now is an indentation issue, here.

File: Polyhedron/demo/Polyhedron/Scene_polyhedron_item.cpp [hunk 11]
```diff
@@ -1835,3 +1843,15 @@ void Scene_polyhedron_item::set_flat_disabled(bool b)
...
+
+void Scene_polyhedron_item::itemAboutToBeDestroyed(Scene_item *item)
+{
+  Scene_item::itemAboutToBeDestroyed(item);
+  if(d && item == this)
+  {
+    delete d->poly;
+    delete d->targeted_id;
+    delete d;
+    d=NULL;
```
Comment (lrineau) [inline Polyhedron/demo/Polyhedron/Scene_polyhedron_item.cpp]: Those lines are strange... the destructor of `d` does not destroy `poly` and `targeted_id`?

**Comment by lrineau** [comment]
This PR adds new warnings with clang and C++14:
```
test/Polyhedron_Demo/Scene_polyhedron_item.h:57:10: warning: 'has_stats' overrides a member function but is not marked 'override' [-Winconsistent-missing-override]
    bool has_stats()const {return true;}
         ^
```
and others. See 
https://cgal.geometryfactory.com/CGAL/testsuite/CGAL-4.10-Ic-166/Polyhedron_Demo/TestReport_lrineau_ArchLinux-clang-CXX14.gz

**Comment by lrineau** [comment]
The PR still triggers warnings about missing `Q_DECL_OVERRIDE` but I will merge it in `master` anyway. The `Q_DECL_OVERRIDE` issue will have a dedicated Github issue.

**Pull request merged**